### PR TITLE
Support AWS Lambda parent inheritance

### PR DESCRIPTION
**Pull request by DarcyRaynerDD**
# What Does This Do

Fixes a bug with Lambda span generation, when using an abstract base class. 
Also fixes a bug where com.amazonaws.services.lambda.runtime.RequestStreamHandler was not being wrapped correctly, because it has a void return type, which the ExtensionAdvice class wasn't matching properly. I've added a test case for this.

Before:
Screenshot 2023-04-19 at 5 04 24 PM
After:
Screenshot 2023-04-19 at 4 52 58 PM

# Motivation

If a customer creates an Abstract Base Class, which implements the `com.amazonaws.services.lambda.runtime.RequestHandler`, currently our instrumentation won't correctly wrap the `handleRequest` method, leading to broken traces. This changes the instrumentation matcher to find all classes in the hierachy which implement the RequestHandler interface.

# Additional Notes

### CODE REVIEW

File: dd-java-agent/instrumentation/aws-lambda-handler/src/main/java/datadog/trace/instrumentation/aws/v1/lambda/LambdaHandlerInstrumentation.java [hunk 1]
```diff
@@ -17,33 +18,30 @@
...
-      } else {
-        this.instrumentedType = handler.substring(0, split);
-        this.methodName = handler.substring(split + HANDLER_SEPARATOR.length());
-      }
-    }
+  }
+
+  @Override
+  public String hierarchyMarkerType() {
+    return "com.amazonaws.services.lambda.runtime.RequestHandler";
```
Comment (maxday) [inline dd-java-agent/instrumentation/aws-lambda-handler/src/main/java/datadog/trace/instrumentation/aws/v1/lambda/LambdaHandlerInstrumentation.java:39]: Is this breaking streaming support? Do we need to add
`com.amazonaws.services.lambda.runtime.RequestStreamHandler`, see Handler interfaces paragraph here: https://docs.aws.amazon.com/lambda/latest/dg/java-handler.html

Comment (DarcyRaynerDD) [reply dd-java-agent/instrumentation/aws-lambda-handler/src/main/java/datadog/trace/instrumentation/aws/v1/lambda/LambdaHandlerInstrumentation.java:39]: Good point, should be easy enough to fix. I'll add a test for it as well.

**Review by maxday — COMMENTED** [review]
LGTM!

File: dd-java-agent/instrumentation/aws-lambda-handler/src/main/java/datadog/trace/instrumentation/aws/v1/lambda/LambdaHandlerInstrumentation.java [hunk 2]
```diff
@@ -108,8 +100,9 @@ static AgentScope enter(
     static void exit(
         @Origin String method,
         @Enter final AgentScope scope,
-        @Advice.Return final Object result,
+        @Advice.Return(typing = DYNAMIC) final Object result,
```
Comment (DarcyRaynerDD) [inline dd-java-agent/instrumentation/aws-lambda-handler/src/main/java/datadog/trace/instrumentation/aws/v1/lambda/LambdaHandlerInstrumentation.java:103]: https://javadoc.io/static/net.bytebuddy/byte-buddy/1.10.5/net/bytebuddy/asm/Advice.Return.html


> Indicates that the annotated parameter should be mapped to the return value of the instrumented method. If the instrumented method terminates exceptionally, the type's default value is assigned to the parameter, i.e. 0 for numeric types and null for reference types. If the return type is void, the annotated value is null if and only if typing() is set to Assigner.Typing#DYNAMIC.

The "handleRequest" method on RequestStreamHandler has a void return type, which was causing this to silently fail to apply the advice for that type.

Comment (bantonsson) [reply dd-java-agent/instrumentation/aws-lambda-handler/src/main/java/datadog/trace/instrumentation/aws/v1/lambda/LambdaHandlerInstrumentation.java:103]: What am I missing here? Is there one method that returns something and one that doesn't? If there isn't then shouldn't the `@Advice.Return ...` just be removed?

Comment (DarcyRaynerDD) [reply dd-java-agent/instrumentation/aws-lambda-handler/src/main/java/datadog/trace/instrumentation/aws/v1/lambda/LambdaHandlerInstrumentation.java:103]: There are two interfaces, `RequestStreamHandler` and `RequestHandler`. On `RequestHandler` the handleRequest method returns a value, (the interface is generic), and `RequestStreamHandler` returns void. We want to capture the return value for the  RequestHandler case

Comment (bantonsson) [reply dd-java-agent/instrumentation/aws-lambda-handler/src/main/java/datadog/trace/instrumentation/aws/v1/lambda/LambdaHandlerInstrumentation.java:103]: Thanks @DarcyRaynerDD, makes perfect sense then.

**Review by bantonsson — APPROVED** [review]

**Comment by bantonsson** [comment]
@DarcyRaynerDD could you please squash the commits when you merge the PR?

**Pull request merged**